File: 98-validate-binary-search-tree/validate-binary-search-tree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

#include <limits.h>
#include <stdbool.h>

bool validate(struct TreeNode* root, long min, long max) {
    
    if (root == NULL)
        return true;

    // current node must lie between min and max
    if (root->val <= min || root->val >= max)
        return false;

    // check left and right subtree
    return validate(root->left, min, root->val) &&
           validate(root->right, root->val, max);
}

bool isValidBST(struct TreeNode* root) {
    
    return validate(root, LONG_MIN, LONG_MAX);
}

File: 98-validate-binary-search-tree/validate-binary-search-tree.kt
class Solution {

    fun isValidBST(root: TreeNode?): Boolean {
        return validate(root, Long.MIN_VALUE, Long.MAX_VALUE)
    }

    fun validate(node: TreeNode?, min: Long, max: Long): Boolean {

        if (node == null) return true

        if (node.`val` <= min || node.`val` >= max)
            return false

        return validate(node.left, min, node.`val`.toLong()) &&
               validate(node.right, node.`val`.toLong(), max)
    }
}